Feature/level 3

maxtalanov opened the pull request.

maxtalanov committed:
refactor(readme): Убрать все лишнее

maxtalanov committed:
refactor(readme   #02): Добавить описание для заголовка

maxtalanov committed:
feat(info#3): Информация о проекте

maxtalanov committed:
feat(skills #4): Описать стек технологий в проекте

maxtalanov committed:
feat(install #5): Правила для локальной установки проекта

maxtalanov committed:
feat(info #6): Дополнить инфо блоком Pull

maxtalanov committed:
feat(full#7): Украсить блоки смайлами

maxtalanov committed:
feat(full#8): Украсить блоки смайлами

maxtalanov committed:
Revert "Feat/movies"

maxtalanov committed:
Merge pull request #8 from maxtalanov/revert-7-feat/movies

Revert 7 feat/movies

maxtalanov committed:
feat(component #1): Добавить индексный фаил

maxtalanov committed:
refactor(up ): auto upadate

maxtalanov committed:
refactor(index #2): Добавить в индексацию компонент APP

maxtalanov committed:
refactor(app #3): Ф-ил стилей импортировать в самом низу

maxtalanov committed:
refactor(app #4): Изменить модель импорта routers

maxtalanov committed:
refactor(app #5): Скорректировать импорты роутинга в каталоге routes

maxtalanov committed:
refactor(app #6): В компоненте app изменить импорты на alias

  Login,
  Main,
  Movies,
  NotFound,
  ProtectedRoute,
  Profile,
  Register,
  SavedMovies

maxtalanov committed:
refactor(index #7): Добавить в индексацию новые компоненты

  Login,
  Main,
  Movies,
  NotFound,
  ProtectedRoute,
  Profile,
  Register,
  SavedMovies

maxtalanov committed:
refactor(alias components): Использовать alias для импорта компонентов

maxtalanov committed:
refactor(alias hooks): Переместить глобальные методы  hook

1.  useFormWithValidation
2. useTimeFormar

maxtalanov committed:
refactor(global metods): Удалить каталог с файлами globalMetods

maxtalanov committed:
refactor(hooks): Добавить новые хуки в индексацию

maxtalanov committed:
refactor(alias hooks): Добавить импорты hook из алиас

maxtalanov committed:
refactor(save movies): Удалить не востребованную ф-ию

maxtalanov committed:
...

maxtalanov committed:
Merge branch 'feat/alias' into feature/level-3

maxtalanov committed:
Востановление...

maxtalanov committed:
Merge branch 'Refactor/README' into feature/level-3

maxtalanov committed:
refactor(readme md): Добавить новый раздел pull-reqvest level3 v2

В виду ошибки мерджа были потеряны коммиты и пул реквест. Из -за этого был создан новый с префиксом v.2

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [ ] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске на странице фильмов текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными. 
    **Комментарий (пояснение, как должен быть реализован функционал повторного показа результата последнего поиска на странице фильмов): для того, чтобы выполнить вышележащий пункт чек-листа, при каждом поиске на странице фильмов результаты поиска нужно сохранять в localstorage. В сохраняемые данные должны входить и строка поиска и статус чекбокса «Короткометражки», чтобы пользователю при повторном переходе на страницу фильмов, можно было вывести на экран информацию, полностью совпадающую с его последним поиском  перед уходом на другую страницу или перезагрузкой страницы (с расстановкой актуального состояния лайков). То есть, сохранённую информацию нужно будет доставать из хранилища при монтировании и перерисовке компоненты фильмов и выводить на экран. При логауте пользователя необходимо эту информацию в хранилище  и стейт-переменных из useState уничтожать. Сейчас пользователь не видит результата своего последнего поиска на странице фильмов при повторном переходе на эту страницу с другой страницы и после перезагрузки сайта - видит отображение всех фильмов без соответствия параметрам поиска.**
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
	**Комментарий: При вводе в поле email адреса почты без точки перед именем домена верхнего уровня (в правой части адреса) можно выполнить запрос на сервер. Однако, сервер возвращает статус ошибки и входа не происходит. Нужно, чтобы валидация имени, почты и пароля и на серверной и на клиентской частях приложения, была одинаковой (по одним и тем же регулярным выражениям), и при невалидных данных запрос на сервер вообще не возможно было отправить.**
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    **Комментарий: Если запрос прошёл успешно, то пользователь так и остаётся на страние регистрации. Зайти на сайт после этого можно только, залогинившись на странице авторизации, а нужен автоматический вход на `/movies` как авторизованного пользователя, без необходимости посещения страницы авторизации.**
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
    **Комментарий: Если запрос прошёл успешно, происходит редирект на главную страницу, а не на `/movies`.**
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
    **Комментарий: Пользователю не отображается уведомление об успешном запросе к серверу при сохранении профиля.**
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.	
    **Комментарий: пользователь может сохранить текущие данные профиля, не изменяя их, а нужно сделать так, чтобы можно было отправить запрос на сохранение, только если  данные изменены по сравнению с текущими данными пользователя, которые должны содержаться в текущий момент в контексте.**
    
    **Напоминание: после каждого успешного сохранения данных профиля на сервере, нужно эти данные вносить в контекст юзера.**
    
    **Замечание: Чтобы данные профиля считались изменёнными, пользователю должно быть достаточно изменения хотя бы одного данного. Может быть изменён один любой инпут из двух, и отправлен запрос на сервер.** 
  - Поиск фильмов:
  **Комментарий: поиск на страницах «Фильмы» и «Сохранённые фильмы» работает не стабильно. После первого поиска, чтобы осуществить следующий, нужна перезагрузка сайта.**
  
   	**Комментарий: Поиск происходит по полному слову. Такое ограничение по поиску очень неудобно для пользователя. Нужно, то, что введено в форме поиска (без пробелов в начале и в конце ввода), рассматривать как подстроку в названии фильма. Если такая подстрока не находится в названии ни одного фильма, должна выводиться надпись "Ничего не найдено", иначе должны выводиться результатаы поиска. Также нужно учесть, чтобы поиск был регистронезависимый.**
    - Прелоадер крутится во время выполнения запроса фильмов.
    **Комментарий: Прелоадер не крутится во время выполнения запроса фильмов.**
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
	**Комментарий: Если ничего не найдено, не выводится надпись «Ничего не найдено» ни на одной из страниц.** 
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  **Комментарий: кнопка «Ещё» остаётся на экране, когда выведен весь результат поиска (см. снимки).**
![Снимок экрана в 2023-04-20 11-15-55](https://user-images.githubusercontent.com/19558891/233292550-95d1595e-ba83-43db-b02c-785ce40d7a55.png)
![Снимок экрана в 2023-04-20 11-20-01](https://user-images.githubusercontent.com/19558891/233292555-bf453910-fd67-4360-b065-460d580dc195.png)
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
    **Комментарий: Клик по карточке не ведёт на трейлер фильма на странице «Сохранённые фильмы».**
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    **Комментарий: Для нового пользователя отображаются активные лайки и карточки, которые сохраняли все другие пользователи. Но, из постановок задач в описании диплома на этапе реакта вытекает, что каждый пользователь должен иметь возможность видеть только свои сохранённые карточки и работать только со своими сохранёнными карточками. То есть, на странице «Сохранённые фильмы» должны присутствовать карточки, избранные только текущим пользователем. Также и на странице фильмов активный статус должны иметь кнопки лайка на карточках, которые добавлены в избранные только текущим пользователем, на всех других карточках кнопки лайка должны иметь неактивный статус.**
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    **Комментарий: при входе на страницу «Сохранённые фильмы» или при поиске на ней должны показываться сразу все фильмы, или все фильмы из результата поиска, без использования кнопки "Ещё".**
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
     **Комментарий: При нажатии на кнопку удаления и при клике на активную кнопку «Лайк» многих карточек, сервер приложения присылает ошибку 403 в ответе (см. снимок вкладки "Сеть" инструментов разработчика в браузере). Возможно, это происходит потому, что в Вашем коде (скорее всего коде бэкенда) ошибочно запрограммировано, что, если какая-то карточка сохранена одним пользователем, её нельзя удалить другому пользователю из своих избранных. Нужно учесть, что разным пользователям совсем не возбраняется добавлять в сохранённые одни и те же фильмы. Хотя, каждый пользователь должен иметь возможность видеть только свои сохранённые карточки и активные лайки, и работать только со своими сохранёнными карточками, в том числе удалять их из своих избранных, что никак не должно влиять на избранные карточки другого пользователя, так как информация о сохранённых карточках на сервере приложения должна храниться раздельно для каждого пользователя.** 
![Снимок экрана в 2023-04-20 11-44-50](https://user-images.githubusercontent.com/19558891/233292764-c01c9c20-7bf5-43a3-99e5-7df65d97de40.png) 
     **Комментарий: При клике на кнопку удаления карточки на странице «Сохранённые фильмы», карточка должна сразу исчезать с экрана на странице «Сохранённые фильмы», а кнопка лайка на этой карточке на странице фильмов должна становиться неактивной. Сейчас этого сразу не происходит, как должно быть. Карточка исчезает со страницы «Сохранённые фильмы» только после перезагрузки сайта. Речь идёт о карточке, которую сохранял текущий пользователь, другие вообще не удаляются (см. предыдущий комм).**
- [ ] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
   **Комментарий: то, что у пользователя накапливаются активные лайки и карточки, которые сохраняли все другие пользователи, ломает пользовательский опыт.**
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [ ] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  **Комментарий: При попытке перейти на любой защищённый роут происходит редирект на страницу авторизации/, а не на `/`.**
  - [ ] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  **Комментарий: При попытке авторизованного пользователя войти на любую страницу по её URL-адресу в адресной строке, происходит вход пользователя на главную страницу, а не на эти страницы, хотя редирект не по набранному адресу должен происходить только при попытке авторизованного пользователя выйти на страницы авторизации и регистрации.**
  - [ ] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  **Комментарий: После успешного вызова обработчика `onSignOut` происходит редирект на страницу авторизации/, а не на `/`.**
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [ ] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [ ] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
- [ ] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 

### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.

## Количество баллов: 42.
**P.S.**
**Все исправления кода должны быть видны в коммитах в репозитории. На исправление каждой ошибки должен быть отдельный коммит. Сделайте, пожалуйста, у каждого коммита понятное название, говорящее, что коммит исправляет. Это обязательные требования.**

maxtalanov committed:
fix(secured route): Исправить маршрут при переходе на защищённый маршрут

Редирекция на страницу "Домашняя или главная".  Маршрут на страницу "/"

Для роутеров (не авторизован):
1.  /movies,
2. /saved-movies,
3.  /profile

maxtalanov committed:
fix(exit route): Исправить редерекцию при выходе из учетки

Пользователь который разогнулся должен попасть на маршрут авторизации

maxtalanov committed:
fix(login route): После успешной регистрации вызвать метод авторизации

После регистрации нового пользователя, при успешной обработки метода, пользователь должен быть авторизован и переадресован на маршрут "/movies"

maxtalanov committed:
feat(profile btn): Добавить стили для заблокированной кнопки

maxtalanov committed:
feat(profile btn): Написать ф-ал для блокировки кнопки отправки

Кнопка отправки должна блокироваться при следующих условиях:
1. Gjk,pjdfntkm bpvtybk jlbПользователь изменил один или несколько инпутов
2. Хотя бы одно поле отличается от входного параметра
3.  Если поле изменено, но в последствии возращено к первоначальному состоянию - кнопка блокируется

maxtalanov commented:
@irinamozes 

> Ваш комментарий:
> Замечание: Чтобы данные профиля считались изменёнными, пользователю должно быть достаточно изменения хотя бы одного данного. Может быть изменён один любой инпут из двух, и отправлен запрос на сервер.

Не совсем понимаю, что вы имеете в виду! 

Алгоритм работы на данный момент: 
1.  Добавлена логика блокировки кнопки (в зависимости от условий);
2.  Можно изменить одно или два поля. После чего будет вызван fetch-запрос по клику редактировать;
3.  Если сервер ответил успех, буде вызван диспач акшина userContext;
4. По итогу будет измене глобальный пользователь и эти данные обновиться во-всех используемых компонентах;
5.Так же, благодаря useEffect, будет обновлен блокировщик кнопки по новым входным данным из контекста. 

Ранее все работало подобным образом, только добавился блокировщик кнопки. По этому не понимаю, что вы мне хотели сказать данным комментарием. 

Заранее спасибо за ответ.

maxtalanov commented:
@irinamozes 

> Ваш комментарий:
> Клик по карточке не ведёт на трейлер фильма на странице «Сохранённые фильмы».

Да действительно, есть фильмы по клику на баннер которого попадаешь на страницу  в ютуб, но ссылка не действительна. Тут проблема уже на стороне API-Яндекса, так как его ссылка не работает, но не на всех фильмах. 

**Как пример фильм под названием _Фавела на взрыве_ - отдает корректную ссылку, а вот фильм _«Роллинг Стоунз» в изгнании_ - отдает битую. Обратите пожалуйста на это внимание при следующей проверки**

maxtalanov committed:
fix(input email): Переписать валидацию для поля инпут email

Нужно перенести patternв регулярное выражение, что валидатор корректно валидировал email адрес

maxtalanov committed:
fix(btn more): Кнопка еще должна прятаться, если отрисованы все фильмы

maxtalanov commented:
> Комментарий: При нажатии на кнопку удаления и при клике на активную кнопку «Лайк» многих карточек, сервер приложения присылает ошибку 403 в ответе (см. снимок вкладки "Сеть" инструментов разработчика в браузере). Возможно, это происходит потому, что в Вашем коде (скорее всего коде бэкенда) ошибочно запрограммировано, что, если какая-то карточка сохранена одним пользователем, её нельзя удалить другому пользователю из своих избранных. Нужно учесть, что разным пользователям совсем не возбраняется добавлять в сохранённые одни и те же фильмы. Хотя, каждый пользователь должен иметь возможность видеть только свои сохранённые карточки и активные лайки, и работать только со своими сохранёнными карточками, в том числе удалять их из своих избранных, что никак не должно влиять на избранные карточки другого пользователя, так как информация о сохранённых карточках на сервере приложения должна храниться раздельно для каждого пользователя.

**UPD:** обновлено на бэке поиск сохраненных фильмов по id текущего пользователя. 
[ссылка на комит](https://github.com/maxtalanov/movies-explorer-api/commit/ae3d579a5a2213165ed10e166ca3e6ffcd21b9b1)

maxtalanov committed:
refactor(save movies): Решить проблему с рендором карточек при удалении

Проблема связана с ф-ал поиском. Части кода которые ломают логику
 - временно закомментировать

maxtalanov committed:
fix(serch movie): Удалить реализованный ранее  ф-ал поиска

maxtalanov committed:
fix(movies): Исправить ошибку отрисовки фильмов

Передать корректный массив

maxtalanov commented:
@irinamozes 
Не совсем понимаю как правильно должен работать филтр поиска, буду признателен за обратную связь в данном плане, если в ниже переданных коммитах сделал не корректно. 
1. Фильтр работает глобально для страницы Фильмов и Сохранённых фильмов, т.е. если поиск осуществлен с одной из страниц, то он дублирует его и на другой или же раздельно - отдельно для фильмоd и сохранённых фильмов? 
2. Свич и инпут работаю синхронно или нет? Т.е. сейчас поиск переделан, на модель по кнопке поиска или же всё-таки свичер, должен работать как отдельная сущность?  Пример: если меняется состояние свича, то автоматом происходит перерисовка, в не зависимости от инпута с кнопкой поиска.

А так буду признателен за пример. Заранее спасибо!

maxtalanov committed:
refactor(hook delete): Удалить hook useStateCash

maxtalanov committed:
feat(hook new): Добавить новый hook useLocalStorage

1. Описать новый хук в отдельном каталоге
2. Добавить hook в индексный фаил

maxtalanov committed:
fix(then movie): Исправить ошибку в блоке then myMovies

1. Добавить заметку для дальнейшего рефакторинга;
2. Скорректировать возвращаемый массив от сервера. Для него нужно добавить новый ключ "isHortFilm" - это необходимо для корректной работы поиска.

maxtalanov committed:
refactor(app): Импортировать все необходимое

maxtalanov committed:
refactor(app): Скорректировать и добавить переменные

maxtalanov committed:
refactor(seach form): Импорты

Импортировать все необходимое и скорректировать импорты

maxtalanov committed:
refactor(seach form): Деструктурировать пропсы и инициализировать форм

maxtalanov committed:
fix(seach form): Удалить все излишнее

maxtalanov committed:
fix(seach form): Скорректировать обработчик submit

maxtalanov committed:
fix(seach form): Скорректировать jsx checkbox, передать корректные парам

maxtalanov committed:
fix(movies): Переработать модель поиска и отрисовки компоненты

irinamozes commented:
@maxtalanov 
Здравствуйте, Максим!

Извините, что не сразу увидела Ваше сообщения.

Отвечаю на Ваши вопросы.

> Замечание: Чтобы данные профиля считались изменёнными, пользователю должно быть достаточно изменения хотя бы одного данного. Может быть изменён один любой инпут из двух, и отправлен запрос на сервер.

Замечание - это не указание на ошибку, а напоминание, что нужно иметь в виду при программировании функционала, или  при внесении изменений в код.

Ответы на последние вопросы.

1. Поиски на страницах Фильмов и Сохранённых фильмов конечно же должны работать раздельно.   Функционалы поиска на страницах «Фильмы» и «Сохранённые фильмы» никак не должны быть связаны, они должны быть полностью независимы друг от друга. Фильтрация по короткометражкам относится к функционалу поиска. Поэтому, если пользователь нашёл короткометражки на странице фильмов, это никаким образом не должно влиять на то, что показывается на странице «Сохранённые фильмы», и наоборот. К тому же,  если у пользователя есть сохранённые фильмы, при переходе на страницу «Сохранённые фильмы» надо выводить на экран все сохранённые карточки, так как пользователь может не помнить, какие фильмы он сохранял, и не найти их.  Нужно, чтобы при входе на эту страницу пользователь видел все свои избранные фильмы без задания параметров поиска, а  не пустой экран и не результат какого-либо поиска.

2. > сейчас поиск переделан на модель по кнопке поиска 

- такой вариант поиска будет принят. Вывод результата поиска должен соответствовать присутствующим на экране параметрам поиска - строке поиска в форме поиска и положению чекбокса "Короткометражки". Можно сделать вывод результатов, только когда пользователь нажмёт кнопку "Найти", на обеих страницах. 

maxtalanov committed:
refactor(stored app): Разделить поиск на два хранилища

maxtalanov committed:
fix(app): Удалить не востребованный  технический код

maxtalanov committed:
refactor(protected router): Удалить комментарий из кода

maxtalanov committed:
refactor(protected router): Деструктурировать пропы отвечающие за auth

maxtalanov committed:
refactor(protected router): Изменить допуск к защитному роуту

 Сделать промежуточный слой проверки для токена пока грузятся данные. После завершения проверки токена - дать разрешение на проверку аутентификации к защищенному роуту и дальнейшей переадресации, если доступа к роуту нет

maxtalanov committed:
refactor(app): Добавить стейт переменную isTokenCheck

maxtalanov committed:
refactor(app): Удалить технический мусор

maxtalanov committed:
refactor(app): Прокинуть пропс  isTokenCheck в протект

maxtalanov committed:
refactor(main): Прокинуть пропс  isTokenCheck и добавить проверку

maxtalanov committed:
fix(app): Исправить  баг с проверкой токена, для не авторизованного user

maxtalanov committed:
feat(app): Добавить индикатор загрузки, на нужные события UI

maxtalanov committed:
Убрать данную реализацию

maxtalanov committed:
style(preloder): Добавить стили для прелодер

maxtalanov committed:
refactor(preloder): Добавить входящий сообщение

maxtalanov committed:
refactor(protected): Добавить прелоуде

maxtalanov committed:
feat(notification): Добавить иконки для компоненты notification

1. Иконка ошибки
2. Иконка информации
3. Иконка успех
4. Иконка закрытия

maxtalanov committed:
feat(index component): Добавить компонент notification в индексный ф-ил

maxtalanov committed:
style(notification): Стилизовать notifocation

maxtalanov committed:
style(notification): добавить компонент и его логику

maxtalanov committed:
refactor(app): Удалить технический мусор

maxtalanov committed:
feat(notification app): Добавить отработку уведомлений

maxtalanov committed:
feat(notification card): Удалить ф-ию auto delete

maxtalanov committed:
refactor(const api): Исправить реджект ошибки

maxtalanov commented:
@irinamozes спасибо за ответы. Также когда будет делать ревью. Буду рад видеть обратную связь по рефакторингу на будущее

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [ ] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске на странице фильмов текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными.
    **Комментарий: Как было написано в комментарии от предыдущей проверки, при логауте пользователя нужно всю информацию вышедшего пользователя в локальном хранилище и также обязательно в стейт-переменных из useState уничтожать. Иначе, информация  пользователя, сделавшего логаут, может стать известна вошедшему следом другому пользователю, не говоря о том, что и сам проект в некоторых случаях будет работать неправильно. Сейчас  следующий пользователь при входе на сайт на странице фильмов видит поиск предыдущего пользователя, а на странице «Сохранённые фильмы» - параметры поиска предыдущего пользователя.**
     
    **Комментарий (пояснение, как должен быть реализован функционал повторного показа результата последнего поиска на странице фильмов): для того, чтобы выполнить вышележащий пункт чек-листа, при каждом поиске на странице фильмов результаты поиска нужно сохранять в localstorage. В сохраняемые данные должны входить и строка поиска и статус чекбокса «Короткометражки», чтобы пользователю при повторном переходе на страницу фильмов, можно было вывести на экран информацию, полностью совпадающую с его последним поиском  перед уходом на другую страницу или перезагрузкой страницы (с расстановкой актуального состояния лайков). То есть, сохранённую информацию нужно будет доставать из хранилища при монтировании и перерисовке компоненты фильмов и выводить на экран. При логауте пользователя необходимо эту информацию в хранилище  и стейт-переменных из useState уничтожать.**
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
   **Комментарий: На странице профиля при попытке пользователя сохранить адрес почты, который принадлежит другому пользователю, сервер приложения присылает в ответе ошибку 500. Но, если почта уже была зарегистрирована другим пользователем, сервер приложения должен присылать ошибку 409 в ответе.**
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.	 
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
    **Комментарий: Клик по карточке не ведёт на трейлер фильма на странице «Сохранённые фильмы».**
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [ ] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  **Комментарий: После успешного вызова обработчика `onSignOut` происходит редирект на страницу авторизации/, а не на `/`.**
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [ ] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
  **Комментарий: при отправке запроса к серверу не нужно очищать поля форм аутентификации до положительного ответа сервера, так как у пользователя должна быть возможность исправить данные в случае возврата сервером ошибки, а не вводить их заново. На время выполнения запроса считается хорошей практикой блокировать поля ввода и кнопку отправки формы, чтобы пользователь не мог выполнить новые запросы до завершения предыдущего. Блокировать кнопку отправки формы на время выполнения запроса нужно без очистки полей формы, а в блоке catch метода с запросом к серверу её нужно разблокировать.**
- [ ] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
**Комментарий: длительность короткометражек, количество фильмов, отображаемое при нажатии кнопки "Еще", другие постоянные числовые данные нужно заносить в константы с понятными названиями и помещать в файл констант, чтобы в коде не было "магических" значений.**
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.
## Доп. Комментарии:
**Комментарий: Если у пользователя есть сохранённые фильмы, при переходе на страницу «Сохранённые фильмы» надо выводить на экран все сохранённые карточки, так как пользователь может не помнить, какие фильмы он сохранял, и не найти их.  Нужно, чтобы при входе на эту страницу пользователь видел все свои избранные фильмы без задания параметров поиска, а  не результат какого-либо поиска. Сохранять данные поиска со страницы «Сохранённые фильмы» в localstorage ни к чему.**
## Количество баллов: 55.
4 балла снято за замечание из комментария в разделе "Доп. Комментарии".

maxtalanov committed:
fix(save movies): Убрать кнопку еще и ф-ал относящийся к ней

Все сохраненные карточки должны быть отрисованы при рендере

maxtalanov committed:
refactor(const storage): Инит переменные для localstorage вынести

maxtalanov committed:
refactor(app): Изменить название  ф-ии на более корректное

maxtalanov committed:
refactor(app): Скорректировать импорта и переменные

maxtalanov committed:
fix(link trailer): Исправить ошибку перехода по клику на постер карточки

maxtalanov commented:
@irinamozes, доброе утро! 

> Доп. Комментарии:
Комментарий: Если у пользователя есть сохранённые фильмы, при переходе на страницу «Сохранённые фильмы» надо выводить на экран все сохранённые карточки, так как пользователь может не помнить, какие фильмы он сохранял, и не найти их. Нужно, чтобы при входе на эту страницу пользователь видел все свои избранные фильмы без задания параметров поиска, а не результат какого-либо поиска. **Сохранять данные поиска со страницы «Сохранённые фильмы» в localstorage ни к чему.**


Т.е.  в данном случае, поиск имеет один цикл, при любом уходе, обновлении, сбросе - поиск возвращается к первичному значению, правильно я понимаю?

irinamozes commented:
@maxtalanov 
Здравствуйте, Максим!
Приводить в нужный вид страницу "Сохранённые фильмы" (то есть показывать все избранные карточки) нужно при входе на эту страницу, то есть при переходе с другой страницы и после перезагрузки. При уходе со страницы, или при перезагрузке, на странице может оставаться результат поиска.

maxtalanov committed:
fix(on exit): При выходе пользователя отправлять на роутер /

maxtalanov commented:
Исправление ошибок на backend'e: 

1. Ошибка в коде: Если E-mail который будет изменен уже зарегистрирован - вернуть ошибку 409 [коммит](https://github.com/maxtalanov/movies-explorer-api/commit/6494caf8c1f575ea0331b02a21c48a126eefd24c)
2. Возвратить релевантное сообщение [коммит](https://github.com/maxtalanov/movies-explorer-api/commit/b139217088302fbc69fbceff4e32a860b0332898)

maxtalanov committed:
refactor(prrofile res): Скорректировать тело ответа для email

В случае прихода ошибки с кодом 409, вернуть подсказку для того, чтоб пользователь понял, что указанный email использовать нельзя

maxtalanov committed:
refactor(app login): Скорректировать сообщение с кодом 401

maxtalanov committed:
refactor(login): Отменить очистку формы при клике на submit

maxtalanov committed:
Стейт переменные и занулиение состояний

maxtalanov committed:
fix(app): Удалить не востребованный комментарий из кода

maxtalanov committed:
fix(app): Удалить не востребованный комментарий из кода

maxtalanov committed:
feat(const): Изменить переменную

maxtalanov committed:
feat(const): Добавить переменную Notificatio config

maxtalanov committed:
refactor(app): Перенести ф-ию render id в переменную notification

maxtalanov committed:
refactor(app): Конф-ть настройки notification из переменой

maxtalanov committed:
refactor(app): Длину короткометражного фильма использовать из переменной

maxtalanov committed:
refactor(const): Длину короткометражек поместить в переменную

maxtalanov committed:
refactor(const): Добавить переменную отрисовки карточек

maxtalanov committed:
refactor(use render card): Использовать для настройки переменные

maxtalanov committed:
Мелкие правки

maxtalanov committed:
fix(api): Удалить не востребованный console.log

irinamozes commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--35--> Функциональность проекта из брифа реализована полностью:
  - Общее
    - Все ссылки и кнопки в проекте работают.
    - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должны быть кнопки «Войти» и «Регистрация»; если пользователь залогинился, кнопки исчезают — и появляются кнопки «Фильмы», «Сохранённые фильмы» и «Аккаунт», в том числе и на главной странице.
    - При поиске на странице фильмов текст запроса, найденные фильмы и состояние переключателя короткометражек сохраняются в хранилище. Если пользователь повторно переходит на страницу фильмов, то при монтировании компонента данные достаются из локального хранилища. Страница отображается в соответствии с загруженными из хранилища данными.
    **Комментарий: Как было уже несколько раз написано в комментариях от предыдущих проверок, при логауте пользователя нужно всю информацию вышедшего пользователя в хранилище и также обязательно в стейт-переменных из useState уничтожать. Сейчас, если следующий пользователь авторизуется на сайте без предварительной перезагрузки сайта после логаута предыдущего пользователя, ему при входе на сайт на странице фильмов отображается поиск предыдущего пользователя. Видимо, эта информация не уничтожается в стейт-переменных из какого-то useState при логауте пользователя.**
	- Все формы валидируются и на стороне клиента. Пользователь не может отправить запрос с невалидными данными.
  - Страницы «Регистрация» и «Авторизация»:
    - На странице «Регистрация» клик по кнопке «Зарегистрироваться» отправляет запрос на роут `/signup`, если данные введены корректно. Если запрос прошёл успешно, то автоматически производится вход и редирект на страницу `/movies`.
    - На странице «Авторизация» клик по кнопке «Войти» отправляет запрос на роут `/signin`, если данные введены корректно. Если запрос прошёл успешно, происходит редирект на страницу `/movies`.
	- Все формы валидируются и на стороне клиента.
  - Страница редактирование профиля:
    - На странице редактирования профиля клик по кнопке «Сохранить» отправляет запрос на роут `/users/me`, если данные введены корректно.
    - Пользователю отображается уведомление об успешном запросе к серверу при сохранении профиля.
	- Если на странице редактирования профиля введённая информация соответствует текущим данным пользователя, кнопка «Сохранить» заблокирована и нельзя отправить запрос сохранения.	 
  - Поиск фильмов:
    - Прелоадер крутится во время выполнения запроса фильмов.
    - Работа с фильтром настроена:
	  - Поиск фильмов регистронезависимый.
	  - Если запрос выполняется впервые, то работа с фильтром происходит после получения данных.
	  - Если карточки уже были отображены на странице в блоке результатов, клик по чекбоксу «Короткометражки» приводит к повторной фильтрации результата.
    - После успешного сабмита формы поиска появляется блок с результатами. 
	- Если ничего не найдено, выводится надпись «Ничего не найдено».
  - На странице всех фильмов в блоке результата отображается такое же количество карточек, как в макете. Нажатие на кнопку «Ещё» отображает следующий ряд с тем же числом карточек. При отображении всех карточек кнопка "Ещё" скрывается.
  - Карточки:
    - Карточка состоит из изображения, названия фильма и его длительности. Длительность фильма рассчитывается корректно и соответствует формату в макете. Клик по карточке ведёт на трейлер фильма.
	- Кнопка лайка имеет правильное состояние, в зависимости от того, добавлен ли фильм в сохранённые или нет.
    - При клике на иконку «Лайк» в блоке карточки выполняется запрос к /movies нашего API для установки или снятия лайка, в зависимости от текущего состояния.
  - На странице «Сохранённые фильмы»:
    - Отображается форма поиска. Она позволяет искать фильмы по уже полученным данным о сохранённых фильмах.
    - Блок карточки содержит кнопку удаления, а не лайка.
    - При нажатии на кнопку удаления выполняется запрос на удаление фильма. После успешного запроса карточка удаляется со страницы.
- [x] <!--8--> Отсутствуют серьёзные баги, которые возникают при работе с сервисом, например: 
  - Удалённые или добавленные карточки фильмов появляются только после перезагрузки страницы.
  - Если осталось отобразить меньше карточек фильмов, чем в полном ряду, то нажатие кнопки «Ещё» вызывает появление ошибок в консоли.
  - При удалении сохранённых карточек на соответствующей странице «Сохранённые фильмы» их по-прежнему можно найти через поиск. Поиск начинает корректно работать только после перезагрузки страницы.
  - Другие баги, которые возникают при работе с сервисом и требуют перезагрузки страницы или «ломают» пользовательский опыт.
- Регистрация и авторизация:
  - [x] <!--2--> Роуты `/saved-movies`, `/movies`, `/profile`  защищёны HOC-компонентом `ProtectedRoute`. Роуты `/` , `/signin`, `/signup` не являются защищёнными. 
  - [x] <!--3--> При попытке перейти на любой защищённый роут происходит редирект на `/` .
  - [x] <!--2--> Если пользователь был авторизован и закрыл вкладку, он может вернуться сразу на любую страницу приложения по URL-адресу, кроме страниц авторизации и регистрации.
  - [x] <!--2--> После успешного вызова обработчика `onSignOut` происходит редирект на `/`.
  - [x] <!--1--> Корректно используется хук `useHistory`.
  - [x] <!--2--> При попытке перейти на несуществующую страницу происходит редирект на страницу «404». Кнопка «Назад» работает корректно.
  - [x] <!--1--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`. 
- Работа с JWT выполнена корректно:
  - [x] <!--2--> JWT-токен хранится в `localStorage` или в `cookie`;
  - [x] <!--2--> `Jwt` проверяется запросом к серверу, а не только в локальном хранилище.
  - [x] <!--3--> При выходе из аккаунта `jwt` удаляется.
- Хуки:
  - [x] <!--1--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--1--> Хуки вызываются в основной функции компонента. 
  - [x] <!--1--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2--> Имена пользовательских хуков начинаются с use.
- Контекст:	
  - [x] <!--1--> В корневом компоненте App создана стейт-переменная, которая хранит данные пользователя. Она используется в качестве значения для провайдера контекста.
  - [x] <!--1--> В компонент `App` внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--1--> Компоненты, в которых используются данные профиля, подписаны на контекст.
- Асинхронные запросы к API:
  - [x] <!--1--> Запросы можно осуществлять через Fetch API или XMLHttpRequest, сторонние библиотеки, такие как axios или jQuery, не применяются.
  - [x] <!--1--> Запросы к API вынесены в отдельные файлы: `MainApi.js` и `MoviesApi.js`.
  - [x] <!--1--> Первый обработчик `then` возвращает `res.json`. res проверяется на корректность.
  - [x] <!--1--> Цепочка обработки промисов завершается блоком catch.
  - [x] <!--2--> Не выполняются лишние запросы к бэкенду, например: запрос всех фильмов с сервиса beatfilm-movies производится только при первом поиске; все сохранённые фильмы не запрашиваются с сервера при каждом лайке или дизлайке; данные пользователя запрашиваются один раз при запуске приложения.
- [x] <!--2--> Именование:
  - имена переменных, функций и параметров написаны в camelCase;
  - имена переменных — существительные;
  - имена переменных, функций и компонентов соответствуют содержимому;
  - имена коллекций NodeList — существительные во множественном числе;
  - имя функции начинается с глагола и отражает то, что она делает;
  - для именования запрещены: транслит и неуместные сокращения.

### Хорошие практики
- [x] <!--1--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--2--> Для элементов списка используется уникальный ключ key, а не индекс массива.
- [x] <!--1--> Запросы к API описаны внутри компонента `App` или в корневых компонентах страниц.
- [x] <!--2--> Для внутренних ссылок в приложении используются компоненты из библиотеки react-router.
- [x] <!--2--> Не происходит утечки памяти при навешивании обработчиков. Все обработчики, добавленные через addEventListener, удаляются при размонтировании компонента.
- Обработка ошибок API:
  - [x] <!--4--> пользователь получает сообщение в случае любой ошибки;
  - [x] <!--2--> поля формы заблокированы во время отправки запросов и у пользователя нет возможности отправить новый запрос до завершения предыдущего.
- [x] <!--2--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
### Рекомендации
- [x] <!--2--> Сторонние JavaScript-библиотеки не применяются.
- [x] <!--1--> Используются семантически правильные блоки для компонентов. Не используются div или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--2--> Отсутствует «мусор» в коде:
  - нет беспорядка в коде;
  - нет лишнего кода: переменная объявлена, но не используется или есть какая-то лишняя логика;
  - код форматирован одинаково, соблюдается иерархия отступов.
## Доп. Комментарии:
**Комментарий: Если у пользователя есть сохранённые фильмы, при переходе на страницу «Сохранённые фильмы» надо выводить на экран все сохранённые карточки. Сейчас все фильмы показываются только после перезагрузки страницы «Сохранённые фильмы», в других случаях при переходе на эту страницу показывается результат последнего поиска, если он был перед уходом.**
## Количество баллов: 94.
2 балла снято за замечание из комментария в разделе "Доп. Комментарии".
4 балла снято за замечание из раздела "Функциональность проекта из брифа реализована полностью". 
## У Вас отличный дипломный проект!

Чтобы красиво оформить файл README.md, где рассказать о собственной работе над проектом и собственных достижениях, можно воспользоваться инструментом Markdown. Описание языка Markdown есть в Ваших лекциях по этой ссылке:
https://praktikum.yandex.ru/learn/web/courses/dbf98e55-0f76-444b-850c-4538708ad571/sprints/1425/topics/b4072eed-2089-45c0-9382-98ea71202341/lessons/939aeab7-0508-49e8-bcf2-199d4dbf74f7/.

Ссылки на интересные материалы и подкасты:
[habr](https://habr.com/ru/)
https://soundcloud.com/everyonecan
https://soundcloud.com/podlodka
https://soundcloud.com/devschacht
https://soundcloud.com/begebot
https://soundcloud.com/web-standards
https://www.youtube.com/channel/UCZeU17nbVfzczAkJVTay9vw

Интересная статья о замыкании и лексическом окружении функций:
https://developer.mozilla.org/ru/docs/Web/JavaScript/Closures.

О пользовательских хуках:
https://ru.reactjs.org/docs/hooks-custom.html

https://reactdev.ru/
Справочник React

Какие бывают стили в названиях файлов js и зачем они нужны, интересно и кратко написано здесь:
https://medium.com/@alivander/camel-pascal-snake-case-и-другие-стили-написания-288ec62ca0d0

Ресурс по библиотеке Redux, которая очень широко используется в реальных проектах на React:
https://rajdee.gitbooks.io/redux-in-russian/content/

Создание приложений JavaScript с помощью TypeScript
https://docs.microsoft.com/ru-ru/learn/paths/build-javascript-applications-typescript/

Руководство Typescript
http://typescript-lang.ru/docs/

TypeScript в деталях.
https://typescript-definitive-guide.ru/book/chapters

**Желаю дальнейших творческих успехов, экспертных знаний и удачи!**

**P.S.**
Рекомендации для дальнейшего развития. Неплохо основательнее изучить программирование на React с использованием библиотеки Redux и надстройки над js TypeScript. Это более совершенный и современный способ программирования, чем использование React без упомянутых инструментов. Все эти инструменты изучаются на курсе React-разработчик. Обычно, те, кто закончил курс Практикума по web разработке, с успехом осваивают и курс  React-разработчик, при этом успешнее, чем те, кто курс по web разработке не кончал.
